### PR TITLE
Fixed issue #19319: potential error fter Updating To LS 6.4.0

### PR DESCRIPTION
**Pull request by Shnoulle**
Dev: use scopes

### CODE REVIEW

**Comment by Shnoulle** [comment]
getSessionKey broken ?

**Comment by Shnoulle** [comment]
Still need update :)

File: application/core/LSUserIdentity.php [hunk 1]
```diff
@@ -67,20 +66,23 @@ public function authenticate()
...
-                    throw new CHttpException(403, gT("You do not have permission to access this page."));
-                }
-                // Delegate actual authentication to plugin
-                $authEvent = new PluginEvent('newUserSession', $this); // TODO: rename the plugin function authenticate()
-                $authEvent->set('identity', $this);
-                App()->getPluginManager()->dispatchEvent($authEvent);
-                $pluginResult = $authEvent->get('result');
-                if ($pluginResult instanceof LSAuthResult) {
-                    $result = $pluginResult;
+                if (!is_null($user) && (!$user->isActive() || $user->isExpired())) {
```
Comment (Shnoulle) [inline application/core/LSUserIdentity.php:72]: An idea can be move this part after plugin event newUserSession : then plugin can update user before test.

For example : 

1. For an LDAP user : check LDAP status `limesureyaccess`
2. For a Saml account : check an attribute 

It's near a new feature, but this fix both issues : didn't allow non-active or expires users + mantis issue 19319

**Comment by Shnoulle** [comment]
OK : tested with plugin (with auto create) + update expire date

File: application/core/LSUserIdentity.php [hunk 2]
```diff
@@ -85,6 +79,16 @@ public function authenticate()
             }
         }
 
+        /* Check user exist, valid and not expired after plugin actions */
+        if ($result->isValid()) {
+            /** @var \User|null */
+            $user = User::model()->notexpired()->active()->findByAttributes(array('users_name' => $this->username));
```
Comment (gabrieljenik) [inline application/core/LSUserIdentity.php:85]: The intention here is to check if the user is ok to login.
Instead of chainning the rules thrugh scopes, wouldn't it be better to:
- Look for the user
- Create a method canLogin()
- Call that method.

That way the method can easily be enriched with more rules.

File: application/models/User.php [hunk 3]
```diff
@@ -996,6 +996,16 @@ public function isExpired()
         return $expired;
     }
 
+    /**
+     * Returns true if user is active
+     * @return boolean
+     */
+    public function isActive()
```
Comment (gabrieljenik) [inline application/models/User.php:1003]: Is this related to this PR?
I believe it is a good method, but asking if related to the solution as it is not called.

Comment (Shnoulle) [reply application/models/User.php:1003]: I add it during devlopement but remove it after.
https://github.com/LimeSurvey/LimeSurvey/pull/3767/commits/1c0cf421b72bad9e7a66fc77c101ea36b0ae3f2f

You're right :)

File: application/core/LSUserIdentity.php (hunk 2 above)
Comment (Shnoulle) [reply application/core/LSUserIdentity.php:85]: :+1:

**Review by gabrieljenik — APPROVED** [review]
